packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -51,6 +51,15 @@ export default function astro({ config, devServer }: AstroPluginOptions): vite.P
       let tsResult: TransformResult | undefined;
       let cssTransformError: Error | undefined;
 
+      // Precheck: verify that the frontmatter is valid TS before sending to our compiler.
+      // The compiler doesn't currently validate the frontmatter, so invalid JS/TS can cause issues
+      // in the final output. This prevents bad JS from becoming a compiler panic or otherwise
+      // breaking the final output.
+      const scannedFrontmatter = /^\-\-\-(.*)^\-\-\-/ms.exec(source);
```

Comment:
@natemoo-re if we can't validate it inside of the compiler, what if the compiler exported a function like `validateFrontmatter` or even `validateJsExpression` so we could run it on template fragments as well? I little nervous about this being async though, if called on every template expression.

Could this be a short-term fix, and then a longer-term fix would be to add this kind of checking into the Go compiler itself later on down the road? We'll probably want the same for valid CSS...

Reply:
If we follow @matthewp’s “only try on catch” strategy then this should only improve the experience. IMO the lexical scan won’t be as important because on catch we know for sure this is invalid JS. Then this would only be for error display and nothing more. So if by some happenstance that RegEx failed, it would only return a weird error message on edge cases, but in most cases would vastly improve the UX.

Reply:
I'm on board with that reasoning and this being the short-term fix. No objections to moving forward with Matthew's suggested approach.